File: src/ts/modelos/cliente.ts
import Documento from "./documento"
import Endereco from "./endereco"
import Telefone from "./telefone"

export default class Cliente {
    private nome: string
    private nomeSocial: string
    private dataNascimento: Date
    private dataCadastro: Date
    private telefones: Telefone[] = []
    private endereco!: Endereco
    private documentos: Documento[] = []
    private dependentes: Cliente[] = []
    private titular!: Cliente

    constructor(nome: string, nomeSocial: string, dataNascimento: Date) {
        this.nome = nome
        this.nomeSocial = nomeSocial
        this.dataNascimento = dataNascimento
        this.dataCadastro = new Date()
    }

    public set Titular(titular: Cliente) { this.titular = titular }
    public set Nome(nome: string) { this.nome = nome }
    public set NomeSocial(nomeSocial: string) { this.nomeSocial = nomeSocial }
    public set DataNascimento(dataNascimento: Date) { this.dataNascimento = dataNascimento }
    public set DataCadastro(dataCadastro: Date) { this.dataCadastro = dataCadastro }
    public set Endereco(endereco: Endereco) { this.endereco = endereco }


    public get Nome() { return this.nome }
    public get NomeSocial() { return this.nomeSocial }
    public get DataNascimento() { return this.dataNascimento }
    public get DataCadastro() { return this.dataCadastro }
    public get Telefones() { return this.telefones }
    public get Endereco() { return this.endereco }
    public get Documentos() { return this.documentos }
    public get Dependentes() { return this.dependentes }
    public get Titular() { return this.titular }

    public removerDependente(dependente: Cliente) {
        let index = this.dependentes.indexOf(dependente)
        if(index >= 0) {
            this.dependentes.splice(index, 1)
        }
    }

    public adicionarDependente(dependente: Cliente) {
        this.dependentes.push(dependente)
    }

    public removerTitularEDepedentes() {
        this.dependentes = []
    }
}

File: backend/src/main/java/com/autobots/automanager/entidades/Cliente.java
package com.autobots.automanager.entidades;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.persistence.*;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import org.springframework.hateoas.RepresentationModel;
import lombok.Data;

@Data
@Entity
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Cliente extends RepresentationModel<Cliente>{

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column
	private String nome;

	@Column
	private String nomeSocial;

	@Column
	private LocalDateTime dataNascimento;

	@Column
	private LocalDateTime dataCadastro;

	@OneToMany(orphanRemoval = true, cascade = CascadeType.ALL)
	private List<Documento> documentos = new ArrayList<>();

	@OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
	private Endereco endereco;

	@ManyToMany
	@JoinTable(
			name = "cliente_telefone",
			joinColumns = @JoinColumn(name = "cliente_id"),
			inverseJoinColumns = @JoinColumn(name = "telefone_id")
	)
	private List<Telefone> telefones = new ArrayList<>();

	@JsonIgnore
	@OneToMany(orphanRemoval = true, cascade = CascadeType.ALL)
	private List<Cliente> dependentes = new ArrayList<>();

	@OneToOne
	private Cliente titular = null;

	@ManyToMany(mappedBy = "hospedes")
	private List<Hospedagem> hospedagens = new ArrayList<>();

	public Cliente(String nome, String nomeSocial, LocalDateTime dataNascimento){
		this.nome = nome;
		this.nomeSocial = nomeSocial;
		this.dataNascimento = dataNascimento;
		this.dataCadastro = LocalDateTime.now();
	}


	public Cliente(){
	}

}